agentendpoint/config_task.go

```diff
@@ -0,0 +1,462 @@
+		asgnmnt := &assignment{policies: map[string]*policy{}}
+		c.assignments[a.GetConfigAssignment()] = asgnmnt
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := &policy{resources: map[string]*resource{}}
+			asgnmnt.policies[p.GetId()] = plcy
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i, r := range p.GetResources() {
+				plcy.resources[r.GetId()] = &resource{}
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[validationStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_Validation{
+						Validation: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation{
+							Outcome: agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_SKIPPED,
+						},
+					},
+				}
+			}
+		}
+	}
+
+	// Unmarshal all resources.
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		ctx = clog.WithLabels(ctx, map[string]string{"config_assignment": a.GetConfigAssignment()})
+		asgnmnt := c.assignments[a.GetConfigAssignment()]
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := asgnmnt.policies[p.GetId()]
+			pResult := aResult.GetOsPolicyResults().GetResults()[i]
+			for i, r := range p.GetResources() {
+				res := plcy.resources[r.GetId()]
+
+				outcome := agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_OK
+				errMsg := ""
+				err := res.unmarshal(ctx, r)
+				if err != nil {
+					outcome = agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation_RESOURCE_PAYLOAD_CONVERSION_ERROR
+					plcy.hasError = true
+					errMsg = fmt.Sprintf("Error unmarshalling resource: %v", err)
+					clog.Errorf(ctx, errMsg)
+				}
+
+				result := pResult.GetResourceResults().GetResults()[i]
+				result.GetExecutionSteps()[validationStepIndex] = &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep{
+					Step: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_ExcecutionStep_Validation{
+						Validation: &agentendpointpb.ApplyConfigTaskOutput_ResourceResult_Validation{
+							Outcome: outcome,
+						},
+					},
+					ErrMsg: errMsg,
+				}
+
+				// Should we continue to validate all resources, or stop on the first unmarshal error?
+				// if plcy.hasError {
+				//   continue
+				// }
+			}
+		}
+	}
+
+	// TODO: check for resouce conflicts
+}
+
+func (c *configTask) checkState(ctx context.Context) {
+	// First populate check state results (for policies that do not have validation errors).
+	for i, a := range c.Task.GetConfig().GetConfigAssignments() {
+		asgnmnt := c.assignments[a.GetConfigAssignment()]
+		aResult := c.results.GetResults()[i]
+		for i, p := range a.GetPolicies() {
+			plcy := asgnmnt.policies[p.GetId()]
+			// Skip state check if this policy already has an error from a previous step.
+			if plcy.hasError {
```

Comment:
I thought that we wanted to fail fast if there is any policy validation error. Did that change?

Reply:
I think we went back and forth on this, I'll double check and update

Reply:
So our current approach is to just fail for this policy, not the entire config run as policies are meant to be separate and not interact with each other.